overlord/snapstate/reboot.go

```diff
@@ -123,6 +174,202 @@ func bootBaseSnapType(byTypeTss map[snap.Type]*state.TaskSet) snap.Type {
 	return snap.TypeOS
 }
 
+// configureSingleRebootForSplitTaskSets sets up restart boundaries for task-sets that have been
+// split into two. It uses the first part of the task-set (i.e the one containing the reboot edges)
+// to set a restart boundary along the 'Do' path for the last task-set, and one along the 'Undo' path
+// for the first task-set, to introduce only one reboot along each direction.
+func configureSingleRebootForSplitTaskSets(byTypeTss, beforeTss map[snap.Type]*state.TaskSet) {
+	// Set reboot boundary along the do path for the last essential
+	for i := len(essentialSnapsRestartOrder) - 1; i >= 0; i-- {
+		o := essentialSnapsRestartOrder[i]
+		if beforeTss[o] == nil {
+			continue
+		}
+
+		// refer to the original task-set that has edges
+		if linkSnap := byTypeTss[o].MaybeEdge(MaybeRebootEdge); linkSnap != nil {
+			restart.MarkTaskAsRestartBoundary(linkSnap, restart.RestartBoundaryDirectionDo)
+		}
+		break
+	}
+
+	// Set reboot boundary along the undo path for the first essential
+	for _, o := range essentialSnapsRestartOrder {
+		if beforeTss[o] == nil {
+			continue
+		}
+
+		// just for consistency lets refer to original task-set
+		if unlinkSnap := findUnlinkTask(byTypeTss[o]); unlinkSnap != nil {
+			restart.MarkTaskAsRestartBoundary(unlinkSnap, restart.RestartBoundaryDirectionUndo)
+		}
+		break
+	}
+}
+
+// ConfigureSnapTaskSetsLinkageAndRestart configures the correct link-order between all
+// the provided snap task-sets, and sets up restart boundaries for essential snaps (base, gadget, kernel).
+// Under normal circumstances link-order that will be configured is:
+// snapd => boot-base (reboot) => gadget (reboot) => kernel (reboot) => bases => apps.
+//
+// However this may be configured into the following if conditions are right for single-reboot
+// snapd => boot-base (up to auto-connect) => kernel (up to auto-connect, then reboot) =>
+// -  boot-base => kernel => bases => apps.
+//
+// XXX: Going forward, gadget will be included in the last example, but this is still TODO.
+// XXX: Going forward, we will also try to move towards reboot-less boot-bases for UC18+
+// as they should in theory be simple enough to not require a reboot.
+func ConfigureSnapTaskSetsLinkageAndRestart(st *state.State, providedDeviceCtx DeviceContext, tss []*state.TaskSet) error {
+	bootBase, err := deviceModelBootBase(st, providedDeviceCtx)
+	if err != nil {
+		return err
+	}
+
+	byTypeTss, err := taskSetsByTypeForEssentialSnaps(tss, bootBase)
+	if err != nil {
+		return err
+	}
+
+	// If the boot-base is 'core', then we don't allow splitting the task-sets to set up
+	// for single-reboot, as we don't support this behavior on UC16.
+	splitAllowed := bootBase != "core"
+	// Attach a separate lane to essential snaps we consider transactional.
+	lane := st.NewLane()
```

Comment:
I think for everything else we want a per-snap lanes, otherwise we get everything else into lane 0 which is I think would be undone together which we don't want?

Reply:
This is handled in snapstate.go:1906 (doUpdate) line, where each snap get their own lane assigned unless `globalFlags.Transaction` is set, where they are all assigned the same lane